File: examples/csharp/v201109/PerformMutateJob.cs
// Copyright 2011, Google Inc. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// Author: api.anash@gmail.com (Anash P. Oommen)

using Google.Api.Ads.AdWords.Lib;
using Google.Api.Ads.AdWords.v201109;

using System;
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System.Threading;

namespace Google.Api.Ads.AdWords.Examples.CSharp.v201109 {
  /// <summary>
  /// This code example shows how to perform asynchronous requests using the
  /// MutateJobService.
  ///
  /// Tags: MutateJobService.mutate, MutateJobService.get
  /// Tags: MutateJobService.getResult
  /// </summary>
  class PerformMutateJob : SampleBase {
    /// <summary>
    /// Returns a description about the code example.
    /// </summary>
    public override string Description {
      get {
        return "This code example shows how to perform asynchronous requests using the" +
            " MutateJobService.";
      }
    }

    /// <summary>
    /// Main method, to run this code example as a standalone application.
    /// </summary>
    /// <param name="args">The command line arguments.</param>
    public static void Main(string[] args) {
      SampleBase codeExample = new PerformMutateJob();
      Console.WriteLine(codeExample.Description);
      codeExample.Run(new AdWordsUser());
    }

    /// <summary>
    /// Run the code example.
    /// </summary>
    /// <param name="user">The AdWords user object running the code example.
    /// </param>
    public override void Run(AdWordsUser user) {
      // Get the MutateJobService.
      MutateJobService mutateJobService = (MutateJobService) user.GetService(
          AdWordsService.v201109.MutateJobService);

      const int RETRY_INTERVAL = 30;
      const int RETRIES_COUNT = 30;
      const int KEYWORD_NUMBER = 100;
      const string INDEX_REGEX = "operations\\[(\\d+)\\].operand";

      long adGroupId = long.Parse(_T("INSERT_ADGROUP_ID_HERE"));

      List<Operation> operations = new List<Operation>();

      // Create an AdGroupAdOperation to add a text ad.
      AdGroupAdOperation adGroupAdOperation = new AdGroupAdOperation();
      adGroupAdOperation.@operator = Operator.ADD;

      TextAd textAd = new TextAd();
      textAd.headline = "Luxury Cruise to Mars";
      textAd.description1 = "Visit the Red Planet in style.";
      textAd.description2 = "Low-gravity fun for everyone!";
      textAd.displayUrl = "www.example.com";
      textAd.url = "http://www.example.com";

      AdGroupAd adGroupAd = new AdGroupAd();
      adGroupAd.adGroupId = adGroupId;
      adGroupAd.ad = textAd;

      adGroupAdOperation.operand = adGroupAd;

      operations.Add(adGroupAdOperation);

      // Create AdGroupCriterionOperation to add keywords.
      for (int i = 0; i < KEYWORD_NUMBER; i++) {
        Keyword keyword = new Keyword();
        keyword.text = string.Format("mars cruise {0}", i);
        keyword.matchType = KeywordMatchType.BROAD;

        BiddableAdGroupCriterion criterion = new BiddableAdGroupCriterion();
        criterion.adGroupId = adGroupId;
        criterion.criterion = keyword;

        AdGroupCriterionOperation adGroupCriterionOperation = new AdGroupCriterionOperation();
        adGroupCriterionOperation.@operator = Operator.ADD;

        adGroupCriterionOperation.operand = criterion;
        operations.Add(adGroupCriterionOperation);
      }

      BulkMutateJobPolicy policy = new BulkMutateJobPolicy();

      policy.prerequisiteJobIds = new long[] {};
      SimpleMutateJob job = mutateJobService.mutate(operations.ToArray(), policy);

      // Wait for the job to complete.
      bool completed = false;
      int retryCount = 0;
      Console.WriteLine("Retrieving job status...");
      while (completed == false && retryCount < RETRIES_COUNT) {
        BulkMutateJobSelector selector = new BulkMutateJobSelector();
        selector.jobIds = new long[] {job.id};

        try {
          Job[] allJobs = mutateJobService.get(selector);
          if (allJobs != null && allJobs.Length > 0) {
            job = (SimpleMutateJob) allJobs[0];
            if (job.status == BasicJobStatus.COMPLETED || job.status == BasicJobStatus.FAILED) {
              completed = true;
              break;
            } else {
              Console.WriteLine("{0}: Current status is {1}, waiting {2} seconds to retry...",
                  retryCount, job.status, RETRY_INTERVAL);
              Thread.Sleep(RETRY_INTERVAL * 1000);
              retryCount++;
            }
          }
        } catch (Exception ex) {
          Console.WriteLine("Failed to fetch bulk mutate job with id = {0}. Exception says \"{1}\"",
              job.id, ex.Message);
          return;
        }
      }

      if (job.status == BasicJobStatus.COMPLETED) {
        BulkMutateJobSelector selector = new BulkMutateJobSelector();
        selector.jobIds = new long[] {job.id};
        JobResult jobResult = mutateJobService.getResult(selector);
        if (jobResult != null) {
          SimpleMutateResult results = (SimpleMutateResult) jobResult.Item;
          if (results != null) {
            if (results.results != null) {
              for (int i = 0; i < results.results.Length; i++) {
                Operand operand = results.results[i];
                Console.WriteLine("Operation {0} - {1}", i, (operand.Item is PlaceHolder) ?
                    "FAILED" : "SUCCEEDED");
              }
            }
            if (results.errors != null) {
              foreach (ApiError error in results.errors) {
                Match match = Regex.Match(error.fieldPath, INDEX_REGEX, RegexOptions.IgnoreCase);
                string index = (match.Success)? match.Groups[1].Value : "???";
                Console.WriteLine("Operation index {0} failed due to reason: '{1}', " +
                    "trigger: '{2}'", index, error.errorString, error.trigger);
              }
            }
          }
        }
        Console.WriteLine("Job completed successfully!");
      } else if (job.status == BasicJobStatus.FAILED) {
        Console.WriteLine("Job failed with reason: " + job.failureReason);
      } else if (job.status == BasicJobStatus.PROCESSING || job.status == BasicJobStatus.PENDING) {
        Console.WriteLine("Job did not complete in {0} secconds.",
            RETRY_INTERVAL * RETRIES_COUNT);
      }
    }
  }
}

File: examples/vb/v201109/PerformMutateJob.vb
' Copyright 2011, Google Inc. All Rights Reserved.
'
' Licensed under the Apache License, Version 2.0 (the "License");
' you may not use this file except in compliance with the License.
' You may obtain a copy of the License at
'
'     http://www.apache.org/licenses/LICENSE-2.0
'
' Unless required by applicable law or agreed to in writing, software
' distributed under the License is distributed on an "AS IS" BASIS,
' WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
' See the License for the specific language governing permissions and
' limitations under the License.

' Author: api.anash@gmail.com (Anash P. Oommen)

Imports Google.Api.Ads.AdWords.Lib
Imports Google.Api.Ads.AdWords.v201109

Imports System
Imports System.Collections.Generic
Imports System.Text.RegularExpressions
Imports System.Threading

Namespace Google.Api.Ads.AdWords.Examples.VB.v201109
  ''' <summary>
  ''' This code example shows how to add ads and keywords using the
  ''' MutateJobService.
  '''
  ''' Tags: MutateJobService.mutate
  ''' </summary>
  Class PerformMutateJob
    Inherits SampleBase
    ''' <summary>
    ''' Returns a description about the code example.
    ''' </summary>
    Public Overrides ReadOnly Property Description() As String
      Get
        Return "This code example shows how to add ads and keywords using the " & _
            "MutateJobService."
      End Get
    End Property

    ''' <summary>
    ''' Main method, to run this code example as a standalone application.
    ''' </summary>
    ''' <param name="args">The command line arguments.</param>
    Public Shared Sub Main(ByVal args As String())
      Dim codeExample As SampleBase = New PerformMutateJob
      Console.WriteLine(codeExample.Description)
      codeExample.Run(New AdWordsUser)
    End Sub

    ''' <summary>
    ''' Run the code example.
    ''' </summary>
    ''' <param name="user">AdWords user running the code example.</param>
    Public Overrides Sub Run(ByVal user As AdWordsUser)
      Dim mutateJobService As MutateJobService = user.GetService( _
          AdWordsService.v201109.MutateJobService)

      Const RETRY_INTERVAL As Integer = 30
      Const RETRIES_COUNT As Integer = 30
      Const KEYWORD_NUMBER As Integer = 100
      Const INDEX_REGEX As String = "operations\[(\d+)\].operand"

      Dim adGroupId As Long = Long.Parse(_T("INSERT_ADGROUP_ID_HERE"))

      Dim operations As New List(Of Operation)

      ' Create an AdGroupAdOperation to add a text ad.
      Dim adGroupAdOperation As New AdGroupAdOperation
      adGroupAdOperation.operator = [Operator].ADD

      Dim textAd As New TextAd
      textAd.headline = "Luxury Cruise to Mars"
      textAd.description1 = "Visit the Red Planet in style."
      textAd.description2 = "Low-gravity fun for everyone!"
      textAd.displayUrl = "www.example.com"
      textAd.url = "http://www.example.com"

      Dim adGroupAd As New AdGroupAd
      adGroupAd.adGroupId = adGroupId
      adGroupAd.ad = textAd
      adGroupAdOperation.operand = adGroupAd
      operations.Add(adGroupAdOperation)

      Dim i As Integer

      For i = 0 To KEYWORD_NUMBER
        Dim keyword As New Keyword
        keyword.text = String.Format("mars cruise {0}", i)
        keyword.matchType = KeywordMatchType.BROAD

        Dim criterion As New BiddableAdGroupCriterion
        criterion.adGroupId = adGroupId
        criterion.criterion = keyword

        Dim adGroupCriterionOperation As New AdGroupCriterionOperation
        adGroupCriterionOperation.operator = [Operator].ADD
        adGroupCriterionOperation.operand = criterion
        operations.Add(adGroupCriterionOperation)
      Next i

      Dim policy As New BulkMutateJobPolicy
      ' You can specify up to 3 job IDs that must successfully complete before
      ' this job can be processed.
      policy.prerequisiteJobIds = New Long() {}

      Dim job As SimpleMutateJob = mutateJobService.mutate(operations.ToArray, policy)

      Dim completed As Boolean = False
      Dim retryCount As Integer = 0
      Console.WriteLine("Retrieving job status...")

      Dim selector As BulkMutateJobSelector
      Do While Not completed
        selector = New BulkMutateJobSelector
        selector.jobIds = New Long() {job.id}
        Try
          Dim allJobs As Job() = mutateJobService.get(Selector)
          If ((Not allJobs Is Nothing) AndAlso (allJobs.Length > 0)) Then
            job = allJobs(0)
            If ((job.status = BasicJobStatus.COMPLETED) OrElse _
                (job.status = BasicJobStatus.FAILED)) Then
              completed = True
              Exit Do
            Else
              Console.WriteLine("{0}: Current status is {1}, waiting {2} seconds to retry...", _
                  retryCount, job.status, RETRY_INTERVAL)
              Thread.Sleep(RETRY_INTERVAL * 1000)
              retryCount = retryCount + 1
            End If
          End If
        Catch ex As Exception
          Console.WriteLine("Failed to fetch bulk mutate job with id = {0}. Exception " & _
              "says ""{1}"".", job.id, ex.Message)
          Return
        End Try
      Loop
      If (job.status = BasicJobStatus.COMPLETED) Then
        selector = New BulkMutateJobSelector
        selector.jobIds = New Long() {job.id}
        Dim jobResult As JobResult = mutateJobService.getResult(selector)
        If Not jobResult Is Nothing Then
          Dim results As SimpleMutateResult = jobResult.Item
          If Not results Is Nothing Then
            If Not results.results Is Nothing Then
              For i = 0 To results.results.Length - 1
                Dim operand As Operand = results.results(i)
                Dim status As String
                If TypeOf operand.Item Is PlaceHolder Then
                  status = "FAILED"
                Else
                  status = "SUCCEEDED"
                End If
                Console.WriteLine("Operation {0} - {1}", i, status)
              Next i
            End If

            If Not results.errors Is Nothing Then
              For Each apiError As ApiError In results.errors
                Dim match As Match = Regex.Match(apiError.fieldPath, INDEX_REGEX, _
                    RegexOptions.IgnoreCase)
                Dim index As String = "???"
                If (match.Success) Then
                  index = match.Groups(1).Value
                End If
                Console.WriteLine("Operation index {0} failed due to reason: '{1}', " & _
                    "trigger: '{2}'", index, apiError.errorString, apiError.trigger)
              Next
            End If
          End If
        End If
        Console.WriteLine("Job completed successfully!")
      ElseIf (job.status = BasicJobStatus.FAILED) Then
        Console.WriteLine("Job failed with reason: " & job.failureReason.ToString())
      ElseIf (job.status = BasicJobStatus.PROCESSING OrElse job.status = BasicJobStatus.PENDING) _
          Then
        Console.WriteLine("Job did not complete in {0} secconds.", RETRY_INTERVAL * RETRIES_COUNT)
      End If

    End Sub
  End Class
End Namespace
